File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        
        // Iterative way to perform Preorder Traversal on Binary Tree using Stack.

        List<Integer>preOrder = new ArrayList<>();

        // check if root is null.

        if(root == null) {
            return preOrder;
        }
        
        Stack<TreeNode>stk = new Stack<>();
        stk.push(root);
        while(!stk.isEmpty()) {
            TreeNode top = stk.peek();
            stk.pop();
            preOrder.add(top.val);
            if(top.right != null) {
                stk.push(top.right);
            }
            if(top.left != null) {
                stk.push(top.left);
            }
        }
        return preOrder;
    }
}

File: 0144-binary-tree-preorder-traversal/0144-binary-tree-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        // solving it using the iterative way using stack.
        vector<int> preOrder;
        if(root == nullptr) {
            return preOrder;
        }
        stack<TreeNode*> stk;
        stk.push(root);
        while(stk.size() > 0) {
            TreeNode *top = stk.top();
            preOrder.push_back(top->val);
            stk.pop();
            if(top->right) {
                stk.push(top->right);
            }
            if(top->left){
                stk.push(top->left);
            }
        }
        return preOrder;  
    }
};